Bluetooth: Controller: Support Link Time Optimizations (LTO)

cvinayak opened the pull request:
Support for using Link Time Optimization (LTO) when
building application with open source Bluetooth Low
Energy Controller. This reduces code space usage.

Use Link Time Optimisation (LTO) when building Controller
with Isochronous channels support.

nRF53x SoCs has 256KB flash and using speed optimisation,
which is the default today, overflows flash when adding
new features related to ISO support.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Support Link Time Optimizations (LTO)

Support for using Link Time Optimization (LTO) when
building application with open source Bluetooth Low
Energy Controller. This reduces code space usage.

Current Memory Usage at the time of this PR (hci_ipc):

BT_CTLR_OPTIMIZE_FOR_SPEED:
Memory region         Used Size  Region Size  %age Used
           FLASH:      260112 B       256 KB     99.22%
             RAM:       60136 B        64 KB     91.76%
           SRAM1:           0 B        64 KB      0.00%
        IDT_LIST:           0 B        32 KB      0.00%

BT_CTLR_OPTIMIZE_FOR_SIZE:
Memory region         Used Size  Region Size  %age Used
           FLASH:      232292 B       256 KB     88.61%
             RAM:       60128 B        64 KB     91.75%
           SRAM1:           0 B        64 KB      0.00%
        IDT_LIST:           0 B        32 KB      0.00%

BT_CTLR_OPTIMIZE_FOR_APP_DEFAULT:
Memory region         Used Size  Region Size  %age Used
           FLASH:      232292 B       256 KB     88.61%
             RAM:       60128 B        64 KB     91.75%
           SRAM1:           0 B        64 KB      0.00%
        IDT_LIST:           0 B        32 KB      0.00%

BT_CTLR_LTO:
Memory region         Used Size  Region Size  %age Used
           FLASH:      221484 B       256 KB     84.49%
             RAM:       60120 B        64 KB     91.74%
           SRAM1:           0 B        64 KB      0.00%
        IDT_LIST:           0 B        32 KB      0.00%

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

henrikbrixandersen commented:
Is this really the way we want to go? Certain subsystems enabling LTO by default, other don't?

cvinayak commented:
> Is this really the way we want to go? Certain subsystems enabling LTO by default, other don't?

LTO has been introduced recently and not enabled for default across the builds probably for a reason? Atleast, for the Bluetooth Controller included builds, i am not observing any regression in my (may be limited) manual testing.

Suggestions welcome if we want all builds to be Link Time Optimized.

henrikbrixandersen reviewed (changes requested):
I think it's too early to automatically enable LTO, even for a subsystem. We are still seeing issues like https://github.com/zephyrproject-rtos/zephyr/pull/70305

nordicjm commented:
> Is this really the way we want to go? Certain subsystems enabling LTO by default, other don't?

Users can enable the setting if they want it manually, I don't really see the value in having it automatically set. Which, I would assume, would then cause problems if they use a debug build and try to step through the bluetooth code

tejlmand commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -583,10 +583,21 @@ config BT_CTLR_DYNAMIC_INTERRUPTS
 	  permit use of SoC's peripheral for custom use when Bluetooth is not
 	  enabled.
 
+config BT_CTLR_LTO
+	bool "Link Time Optimization"
```
This is not a good location for this setting.

It may give the impression that LTO is only applied on this subsystem, which is actually not the case.
Enabling LTO here will enable LTO for the Zephyr build.

One may say, `of course, that should be obvious`, but in Zephyr we are actually having parts which could be using LTO while other parts would not be using it, for example for the TF-M secure build.

cvinayak replied:
I have now added `depends on ARM && BT_HCI_RAW`, this will restrict LTO to samples/application that build only the controller subsystem (aka hci_uart, hci_ipc... and so on) with a HCI transport.

And on the hci_ipc for now needs reduced code space usage in-order to support ISO features and be able to have spare space for DFU.

cvinayak replied:
Removed `&& BT_HCI_RAW`  as normal samples/application would like to use LTO.

tejlmand reviewed (changes requested):
> I think it's too early to automatically enable LTO, even for a subsystem. We are still seeing issues like #70305

I agree, considering the complexity of Zephyr, the need for whole-archive, link post-processing, etc. then I think there are too many variables to enable this per-default.

Although this has been tested for a single sub-system, then what happens when that subsystem is combined with the myriad of other options / subsystems / SoCs which may not have been verified to work properly with LTO ?

Also, having this config option inside a sub-system can give the false impression that the setting only applies to part of the build, which is not the case.

Technically, if using partial linking of subsystems before a final linking it would be possible to apply lto on just part of the linking (gcc/ld) on a subsystem, but that's not what this setting does, and would be a completely different task 🙂 .

cvinayak committed:
samples: Bluetooth: hci_ipc: Use LTO when building ISO features

Use Link Time Optimisation (LTO) when building Controller
with Isochronous channels support.

nRF53x SoCs has 256KB flash and using speed optimisation,
which is the default today, overflows flash when adding
new features related to ISO support.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@henrikbrixandersen  and @tejlmand please re-review.

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -593,12 +593,22 @@ config BT_CTLR_DYNAMIC_INTERRUPTS
+	bool "HCI Controller with Link Time Optimization"
+	depends on ARM
+	select SHARED_INTERRUPTS
+	select ISR_TABLES_LOCAL_DECLARATION
+	select LTO
+	help
+	  Enable use of Link Time Optimizations, to reduce code space usage.
+	  This disables compiler optimization for execution speed.
+
 config BT_CTLR_OPTIMIZE_FOR_SPEED
-	prompt "Optimize for Speed" if !(SOC_SERIES_NRF51X && BT_CTLR_LE_ENC)
-	bool
-	default y if BT_CTLR_LE_ENC
+	bool "Optimize for Speed"
+	depends on !BT_CTLR_LTO
+	default y
 	help
-	  Optimize compilation of controller for execution speed.
+	  Optimize compilation of Controller for execution speed.
```
Given that these two are mutually exclusive, should they have been a `choice`?

cvinayak replied:
Changed to `choice`

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -593,12 +593,22 @@ config BT_CTLR_DYNAMIC_INTERRUPTS
-	default y if BT_CTLR_LE_ENC
+	bool "Optimize for Speed"
+	depends on !BT_CTLR_LTO
+	default y
```
Can you expand the commit message (or add a comment here) on why speed is the default? 
The default for Zephyr is to optimize for size 
```
choice COMPILER_OPTIMIZATIONS
	prompt "Optimization level"
	default NO_OPTIMIZATIONS    if COVERAGE
	default DEBUG_OPTIMIZATIONS if DEBUG
	default SIZE_OPTIMIZATIONS
	help
	  Note that these flags shall only control the compiler
	  optimization level, and that no extra debug code shall be
	  conditionally compiled based on them.
```

cvinayak replied:
Added to help text.

wopu-ot commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -598,12 +598,29 @@ config BT_CTLR_DYNAMIC_INTERRUPTS
 	  enabled.
 
+choice
+	prompt "Optimazation options"
```
Looks like a typo snuck in here: "Optimazation" -> "Optimization"

cvinayak replied:
Thanks, just realized that moving to `choice` means I need to also have the optimize for space which would be the alternative when optimize for speed need to be disabled (and no LTO) for the Controller.

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -598,12 +598,29 @@ config BT_CTLR_DYNAMIC_INTERRUPTS
-	prompt "Optimize for Speed" if !(SOC_SERIES_NRF51X && BT_CTLR_LE_ENC)
-	bool
-	default y if BT_CTLR_LE_ENC
+	bool "Optimize for Speed"
```
Should this depend on `!LTO`? 

Otherwise it is possible to set `CONFIG_BT_CTLR_LTO=n` but `CONFIG_LTO=y`

Similarly, should `BT_CTLR_LTO` depend on `LTO` intead of `select`ing it?

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -598,12 +598,29 @@ config BT_CTLR_DYNAMIC_INTERRUPTS
 
+choice
+	prompt "Optimazation options"
+	default BT_CTLR_OPTIMIZE_FOR_SPEED
```
Should we default to `BT_CTLR_LTO` if `LTO=y`?

cvinayak replied:
Moved out `BT_CTLR_LTO` from the `choice`  to make LTO and compiler optimization mutually exclusive.

Thalley reviewed (commented):
There are some interdependencies that looks messy to me. 

Currently it's possible to set `LTO` and `BT_CTLR_OPTIMIZE_FOR_SPEED`

Thalley commented on tests/bluetooth/init/prj_ctlr_4_0_dbg.conf:
```diff
@@ -25,6 +25,7 @@ CONFIG_BT_CTLR_CONN_RSSI=n
 CONFIG_BT_CTLR_ADV_INDICATION=n
 CONFIG_BT_CTLR_SCAN_REQ_NOTIFY=n
 CONFIG_BT_CTLR_SCAN_REQ_RSSI=n
+CONFIG_BT_CTLR_OPTIMIZE_FOR_APP_DEFAULT=y
```
It's unclear why these specific 3 .conf files have these specific values set. 

Please expand the commit message to explain.

cvinayak replied:
These are added as coverage in CI for newly added KConfigs.... New  Kconfig are generally covered either in samples.yaml or in tests/bluetooth/init  builds.

Thalley replied:
Okay, thanks, but please put that in the commit message :)

cvinayak replied:
added as separate commit.

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -598,12 +598,44 @@ config BT_CTLR_DYNAMIC_INTERRUPTS
+choice
+	prompt "Optimization options"
+	depends on !LTO
+	default BT_CTLR_OPTIMIZE_FOR_SPEED
```
Previously we defaulted to `default y if BT_CTLR_LE_ENC` but now it's always the default. Is that intentional?

cvinayak replied:
yes, intentional. This is leftover from legacy LL implementation, this is not a requirement upstream now.

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -598,12 +598,44 @@ config BT_CTLR_DYNAMIC_INTERRUPTS
+
+endchoice
+
+config BT_CTLR_LTO
```
From your commit message it looks like `BT_CTLR_OPTIMIZE_FOR_SIZE` is worse for size than `BT_CTLR_LTO`. 

Is there a downside to `BT_CTLR_LTO` since we have both? Otherwise it would appear as `BT_CTLR_OPTIMIZE_FOR_SIZE` could be removed as just being a worse version of `BT_CTLR_LTO`. 

What does the size comparison look like if you just compared `BT_CTLR_OPTIMIZE_FOR_SIZE` to `LTO`, i.e. have the same values for `SHARED_INTERRUPTS` and `ISR_TABLES_LOCAL_DECLARATION`?

cvinayak replied:
LTO option for gcc compiler does not optimize when generating object files (my understanding for now), the linker finds instruction patterns to reuse code, hence optimization considers across object files (copy-paste code across c files sometimes lead to lower code space with LTO!)

LTO gives ~4% more optimization compared to `-Os` (compiler optimization for size) in the Controller code base.

Compiler optimization for Size ver Speed gives ~10%.

For:
```
-CONFIG_BT_CTLR_LTO=y
+CONFIG_BT_CTLR_OPTIMIZE_FOR_SIZE=y
+CONFIG_SHARED_INTERRUPTS=y
+CONFIG_ISR_TABLES_LOCAL_DECLARATION=y
+# CONFIG_BT_CTLR_LTO=y
```

I get:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      233528 B       256 KB     89.08%
             RAM:       60128 B        64 KB     91.75%
           SRAM1:           0 B        64 KB      0.00%
        IDT_LIST:           0 B        32 KB      0.00%
```

i.e  more code added to `-Os` with `SHARED_INTERRUPTS` and `ISR_TABLES_LOCAL_DECLARATION`

Thalley replied:
Thanks. Does it make sense to keep both `BT_CTLR_OPTIMIZE_FOR_SIZE` and `BT_CTLR_LTO` then? Or does `BT_CTLR_LTO` only work for GCC?

cvinayak replied:
for now ` BT_CTLR_LTO` needs `LTO` in Zephyr... but I do not see any restrictions set on use of GCC for `LTO` in Zephyr Kconfig other than `ARM` and around the interrupt table constructs.

Today  by this PR,  `BT_CTLR_LTO` uses the zephyr project default of `-Os` to give the best code reduction. I tried to force `BT_CTLR_LTO` with `-Ofast`, this resulted in 16K Flash overflow. Probably, LTO and `-Ofast`  is broken, hence, I will keep the changes in this PR for now as-is, to not mix compiler optimiaztion with link time optimization.

cvinayak committed:
Bluetooth: Controller: Remove legacy LL optimize for speed dependency

Remove legacy LL optimize for speed dependency required to
support encryption feature in Controller where crypto was
performed in highest priority ISRs to setup encryption in 3
radio events (this now requires 5 radio events in split LL).

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: init: Build time coverage for Controller optimization

Added build time coverage for added new Controller
optimization related Kconfigs.

New Kconfigs are either covered in samples.yaml or in the
tests/bluetooth/init conf files.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
> @henrikbrixandersen and @tejlmand please re-review.

@henrikbrixandersen and @tejlmand  Any chance you can take a second look as to if the concerns are addressed?

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -76,6 +76,18 @@ config BT_LLL_VENDOR_OPENISA
+	  This disables compiler optimization for execution speed or space, and
+	  the Link Time Optimizations apply to the whole of the application.
+
+config LTO
+	select BT_CTLR_LTO_SUPPORT
```
Isn't this missing `BT_CTLR` prefix, or are you intentionally overwriting the Kconfig option?

cvinayak replied:
This is intentionally appending selection of Controller's LTO support when application enables `CONFIG_LTO=y`. Thereafter hidden `BT_CTLR_LTO_SUPPORT` with its `depends on` will warn the application to include `SHARED_INTERRUPTS`  et al needed as the dependencies for `CONFIG_LTO` support in Controller.

This change is in response discussion with @tejlmand 

```
[9:36 AM] Rasmussen, Torsten
We have an LTO symbol depending on ISR_TABLES_LOCAL_DECLARATION, and now you add another LTOsymbol which aim is to select ISR_TABLES_LOCAL_DECLARATION. Meaning there are two symbols for enabling LTO.
[9:37 AM] Rasmussen, Torsten
[select recommendations](https://docs.zephyrproject.org/latest/build/kconfig/tips.html#id9)[](https://docs.zephyrproject.org/latest/build/kconfig/tips.html#select-recommendations)
In summary, here are some recommended practices for select:
Avoid selecting symbols with prompts or dependencies. Prefer depends on. If depends on causes annoying bloat in configuration files, consider adding a Kconfig default for the most common value.
 
 
Kconfig - Tips and Best Practices — Zephyr Project Documentation
[9:37 AM] Rasmussen, Torsten
https://docs.zephyrproject.org/latest/build/kconfig/tips.html#select-recommendations
Kconfig - Tips and Best Practices — Zephyr Project Documentation
[9:38 AM] Rasmussen, Torsten
[Using select for helper symbols](https://docs.zephyrproject.org/latest/build/kconfig/tips.html#id8)[](https://docs.zephyrproject.org/latest/build/kconfig/tips.html#using-select-for-helper-symbols)
A good and safe use of select is for setting “helper” symbols that capture some condition. Such helper symbols should preferably have no prompt or dependencies.
 
 
Kconfig - Tips and Best Practices — Zephyr Project Documentation
```

Thalley replied:
Hmm, so you want `CONFIG_LTO` to enable `CONFIG_BT_CTLR_LTO_SUPPORT` if dependencies are enabled, and else give a warning? 

I think adding 
```c
config LTO
	select BT_CTLR_LTO_SUPPORT
```
is even worse than the `select` that @tejlmand didn't like. 

IMO it should be
```
config BT_CTLR_LTO_SUPPORT
	bool
	depends on LTO && ARM && SHARED_INTERRUPTS && ISR_TABLES_LOCAL_DECLARATION
```
instead.

cvinayak replied:
Who selects `BT_CTLR_LTO_SUPPORT` ?

Thalley replied:
> Who selects `BT_CTLR_LTO_SUPPORT` ? i will have to try and get back

What do you mean? I was quoting the code you are adding here

Thalley replied:
Indeed, and I see how that is an issue. 

But it should also be possible to enable `CONFIG_LTO` without getting warnings for `CONFIG_BT_CTLR_LTO_SUPPORT` IMO. 

There may be some Kconfig-magic we can apply to get a good result, but I don't have a good solution right now. I'd like to get some additional eyes on this, just in case that I'm just in the wrong :)
